### ISSUE TITLE
Recommended Speed Improvements

### ISSUE THREAD

**Issue by paulprocacci** [charles-cooper/itch-order-book#10]
Yours:
-----------
\# cpuset -c -l4 /usr/bin/time -h ./a.out < NASDAQ.ITCH50
        31.84s real             7.66s user              23.68s sys

Mine:
-----------
\# cpuset -c -l4 /usr/bin/time -h ./maker < NASDAQ.ITCH50
       19.68s real             6.59s user              13.08s sys
       
Neither program was set to build an actual orderbook so this is just testing the ability to read the packets from the accompanying file and provide them to a given function to process the packets.  My version of this particular part of the program is 38.19% faster than yours.

The difference in speed comes down to you copying the packet contents into `buf', while my program simply returns a pointer to what resides in the network buffer ensuring at all times the complete packet contents exists.

To put it another way, in my version there is no copying from userland buffer to userland buffer.  The copy happens just once when read from the socket.

My version is also written in C but that is mostly irrelevant.  Here is the actual function which handles returning the pointer:

```
void *circular_getp(circular *c, size_t n, int *match)
{
  uint32_t avail;

  if(n > PAGESIZE){
    *match = -1;
    errno = EINVAL;
    return 0;
  }

  while((avail = c->tail - c->head) < n){
    ssize_t r = oneread(c->op,c->fd,c->x + c->tail, PAGESIZE - avail);
    if(r <= 0) { *match = r; return 0; }
    c->tail += r;
  }

  *match = 1;
  return c->x + c->head;
}

```

Avoid copying when you don't have to!

Cheers!

Edit: I couldn't get the code to align properly.  Github hates me.  Sorry.  ;(

**Comment by charles-cooper** [comment]
@paulprocacci looks like the userland portion of the program runs 1s faster; the sys portion is the bulk of the time difference. Is it possible because of the order you ran the tests in, the file was already warmed up in the buffer cache by the time you ran your program?

**Comment by paulprocacci** [comment]
This test is repeatable and has nothing to do with caches.

Here's yours with multiple runs:

\# cpuset -c -l4 /usr/bin/time -h .a.out < NASDAQ.ITCH50
        31.56s real             7.50s user              23.96s sys
\# cpuset -c -l4 /usr/bin/time -h ./a.out < NASDAQ.ITCH50
        31.29s real             7.92s user              23.25s sys
\# cpuset -c -l4 /usr/bin/time -h ./a.out < NASDAQ.ITCH50
        31.24s real             7.37s user              23.86s sys

.. and again mine:        
\# cpuset -c -l4 /usr/bin/time -h ./maker < NASDAQ.ITCH50
        20.16s real             7.01s user              13.10s sys
\# cpuset -c -l4 /usr/bin/time -h ./maker < NASDAQ.ITCH50
        21.13s real             7.12s user              13.60s sys
\# cpuset -c -l4 /usr/bin/time -h ./maker < NASDAQ.ITCH50
        20.67s real             7.32s user              13.34s sys

I do think you're partially right.  My avoidance of copying from userland buffer to userland buffer is saving me a slight bit of time.
Looking closer at the syscalls it's apparent that your program called read(2) 4 times more than my program does.

Your syscall breakdown:
--------------------------
```
syscall                     seconds   calls  errors
madvise                 0.000005881       1       0
write                   0.000023174       4       0
ioctl                   0.000004719       1       0
__sysctl                0.000004288       1       0
readlink                0.000004448       1       1
munmap                  0.000023355       5       0
pread                   0.000005531       1       0
getdirentries           0.000022783       2       0
fcntl                   0.000003807       1       0
close                   0.000039346      10       0
read                   67.95330002612749516       0
fstat                   0.000047801      10       0
open                    0.000058739      11       1
sigfastblock            0.000003627       1       0
issetugid               0.000006763       2       0
mprotect                0.000029854       7       0
mmap                    0.000269560      40       0
                      ------------- ------- -------
                       67.95385370212749614       2
```

My syscall breakdown:
--------------------------
```
syscall                     seconds   calls  errors
ftruncate               0.000004989       1       0
shm_open2               0.000005520       1       0
__sysctl                0.000004238       1       0
readlink                0.000004348       1       1
munmap                  0.000004939       1       0
pread                   0.000005440       1       0
getdirentries           0.000023364       2       0
fcntl                   0.000004168       1       0
close                   0.000024466       6       0
read                   24.634478787 3153335       0
fstat                   0.000021441       5       0
open                    0.000032241       6       0
sigfastblock            0.000004619       1       0
issetugid               0.000006734       2       0
mprotect                0.000021162       5       0
mmap                    0.000087776      14       0
                      ------------- ------- -------
                       24.634734232 3153383       1
```

This would imply to me your buffering strategy can be improved.  I've attached my circular buffer logic which may or may not be useful to you.  Again it's written in C and I'm not familiar enough with C++ to provide you with a version that is suitable for your application.

[circular.c.txt](https://github.com/charles-cooper/itch-order-book/files/9890254/circular.c.txt)
[circular.h.txt](https://github.com/charles-cooper/itch-order-book/files/9890255/circular.h.txt)

A quick explanation of my circular buffer logic is:

- Create a block of shared memory that is PAGESIZE in length.
- MMap a block of memory that is twice the size of the shared memory block:  i.e  PAGESIZE << 1
- Map the shared memory block to the first and second halves of the mmap'd block.

From there, the function circular_getp a) ensures there is enough data in the block of memory returning a pointer to the head of the block and b) circular_advance not only advances but adjusts the head and tail pointer accordingly once we've advanced past 1/2 the mmap'd region.

As a side note, the function circular_get is unused and can be ignored entirely.  It is a remnant of my userland copy to userland copy logic that is no longer used.

Anyways, hope this helps.

**Comment by charles-cooper** [comment]
@paulprocacci not sure how you are comparing these. The existing code doesn't copy between userland buffers; it does do some parsing though and it's unclear if you are counting that or not.

The reason for 4x more reads is that your read size is 4096 bytes, https://github.com/charles-cooper/itch-order-book/blob/095ac97ef46c72411a62dd9148d96d2e3ace3305/main.cpp#L41 uses a buffer size of 1024. There is a tradeoff here; if the buffer size is larger, that eats up more cache. I believe I benchmarked and tuned this at one point but that was probably ten years ago, the results could be different on a more recent CPU.

**Comment by paulprocacci** [comment]
Cool, suit yourself.  ;)
4096 bytes is the standard for nearly every modern operating system, and reading tiny little bits at a time and causing needless context switches is far more time consuming than trashing a cache.

**Comment by charles-cooper** [comment]
@paulprocacci the performance of the order book is extremely sensitive to cache availability. That being said, I am definitely open to benchmarks demonstrating that performance (of the whole system) improves on recent CPUs when this particular buffer increases in size.

**Comment by paulprocacci** [comment]
Don't know what else to say other than test it.  I will too when I have time.

Sure, the read of 1024 bytes destroys 16 L1 cache lines.
The read of 4096 bytes destroy 64 L1 cache lines.

For extremely small data sets, nothing will get removed from L2 and L3 cache lines.  The access times to the L2 and L3 caches are sub 10ns.

Time it takes to perform a context switch is: 5 μs (roughly)
Time it takes to read a block of memory into L1 cache that isn't already present in L2 or L3 cache: 10ns (roughly)

Anyways, I'll put it on my todo to get graphs but please don't hold your breadth.  Cantor Fitzgerald keeps me quite busy.
